README.md
The latest version:
milter-regex 3.1   January 15th, 2023

- - -

milter-regex 3.0   April 23rd, 2022  

New feature: GeoIP  

(1) Everything about GeoIP processing is in the milter-regex source codes.  
GeoIP is implemented without MaxMind Inc.'s libraries or other GeoIP libraries.  
No dependency on the libgeoip.  

(2) Use the RIR ( Regianl Internet Registry ) IP address allocation lists for GeoIP data.  
Get country code by looking at the IP address allocation lists of the RIR.  
Those lists are converted from ASCII format to binary format by using  
the standalone utility program included in the milter-regex package.  

(3) Adding new keyword 'country' and some settings keywords for GeoIP.


ChangeLog
3.1: January 15th, 2023
Output country code to log messages

3.0: April 23rd, 2022
Takao Abe added GeoIP filtering criteria, you can find his version on github.com/milter-regex.

2.7: December 12th, 2019
Add -t option to test the configuration file and exit with a status, suggested by Ralph Seichter.

2.6: April 26th, 2019
Treat socket file name without prefix like local file, from Takao Abe. Make pid file writable
only by root, from Ralph Seichter.

2.5: April 18th, 2019
Add -r option to write pid file. Based on FreeBSD port patches.

2.4: March 2nd, 2019
Add -f option to set syslog facility. Patch from Takao Abe.

2.3: January 28th, 2019
Bug fix: for actions followed by multiple expressions (not just one arbitrarily complex expression),
when multiple expressions become defined during the same sequence point, but with different values
(e.g. one true, another false), depending on the expression order, the action might not be taken,
 when it should be.
This affects all prior versions since 1.0. As a workaround, use only a single expression per action
(duplicating action lines where needed), or combine multiple expressions to a single expression
per action using 'or'.
Report and testing by JCA.

2.2: September 25, 2018
Add -U, -G, and -P options to set pipe user, group, and permissions. Suggested and tested by Ralph Seichter.

2.1: September 26, 2017
Default maximum log level to 6 (LOG_INFO), i.e. exclude LOG_DEBUG.

2.0: November 25, 2013
Add -l option to specify maximum log level.

1.9: November 21, 2011
Add -j option to chroot. Improve building on various platforms. Fix some typos in documentation
 and example config.

1.8: August 12, 2010
Log symbolic host name together with numeric IP address.

1.7: August 4, 2007
Support filtering sendmail macros, like {auth_type}.

1.6: June 6, 2005
Support sendmail quarantine action. Requires non-ancient sendmail (>= 8.13) and libmilter,
as shipping with recent *BSD releases by default.
More fixes for the state machine, dealing with multi-message connections.

1.5: March 19, 2004
Fix logic errors in dealing with multi-message connections (SMTP RSET, HELO or MAIL FROM resetting
SMTP state). Add cb_abort callback.

1.4: March 13, 2004
Some performance improvements, abort rule evaluation immediately when no further rules can possibly match.
 Compile without -Werror, as some ports generate warnings.

1.3: March 8, 2004
Two bugfixes related to RCPT TO: rule evaluation (DSN options and multiple receipients would match
incorrectly), umask(0177) for pipe, fix for Solaris daemon() implementation. Improved logging
(From:, To: and Subject: headers, when available).

1.2: February 27, 2004
Some logging improvements and small fixes. Adds Makefiles for GNU/Linux and Solaris.
Thanks to everyone who helped me solve the build problems.

1.1: February 25, 2004
Support macro definition/expansion.

1.0: February 24, 2004
Now supports boolean expressions, so multiple regular expressions can be combined using
and, or, not and parentheses.
Note that the new parser now requires quotes around reject/tempfail messages.
If you get syntax errors in your existing configuration file, lacking quotes are a likely cause.
Otherwise rulesets are backwards compatible with pre-1.0 versions.

0.1: September 24, 2003
First version.

eval.c
/* $Id: eval.c,v 1.3 2019/01/28 17:12:52 dhartmei Exp $ */

/*
 * Copyright (c) 2004-2019 Daniel Hartmeier
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *    - Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    - Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

static const char rcsid[] = "$Id: eval.c,v 1.3 2019/01/28 17:12:52 dhartmei Exp $";

#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
#include <string.h>

#include "eval.h"

#ifndef REG_BASIC
#define REG_BASIC	0
#endif

extern int	 yyerror(const char *, ...);
extern void	 die(const char *);
static void	 mutex_lock(void);
static void	 mutex_unlock(void);
static int	 check_cond(struct cond *, const char *, const char *);
static void	 push_expr_result(struct expr *, int, int *);
static void	 push_cond_result(struct cond *, int, int *);
static int	 build_regex(struct cond_arg *);
static void	 free_expr_list(struct expr_list *, struct expr *);

static pthread_mutex_t	 mutex;
static struct action	 default_action;

int
eval_init(int type)
{
	memset(&default_action, 0, sizeof(default_action));
	default_action.type = type;
	if (pthread_mutex_init(&mutex, 0))
		return (1);
	else
		return (0);
}

static void
mutex_lock(void)
{
	if (pthread_mutex_lock(&mutex))
		die("pthread_mutex_lock");
}

static void
mutex_unlock(void)
{
	if (pthread_mutex_unlock(&mutex))
		die("pthread_mutex_unlock");
}

struct ruleset *
create_ruleset(void)
{
	struct ruleset *rs;

	rs = calloc(1, sizeof(struct ruleset));
	if (rs == NULL) {
		yyerror("create_ruleset: calloc: %s", strerror(errno));
		return (NULL);
	}
	return (rs);
}

struct expr *
create_cond(struct ruleset *rs, int type, const char *a, const char *b)
{
	struct cond *c = NULL;
	struct cond_list *cl = NULL;
	struct expr *e = NULL;
	struct expr_list *elc = NULL;

	mutex_lock();
	e = calloc(1, sizeof(struct expr));
	if (e == NULL)
		goto error;
	elc = calloc(1, sizeof(struct expr_list));
	if (elc == NULL)
		goto error;

	for (cl = rs->cond[type]; cl != NULL; cl = cl->next) {
		if ((cl->cond->args[0].src == NULL) != (a == NULL) ||
		    (cl->cond->args[1].src == NULL) != (b == NULL) ||
		    (a != NULL && strcmp(a, cl->cond->args[0].src)) ||
		    (b != NULL && strcmp(b, cl->cond->args[1].src)))
			continue;
		break;
	}
	if (cl != NULL)
		c = cl->cond;
	else {
		cl = calloc(1, sizeof(struct cond_list));
		if (cl == NULL)
			goto error;
		c = calloc(1, sizeof(struct cond));
		if (c == NULL)
			goto error;

		if (a != NULL) {
			c->args[0].src = strdup(a);
			if (c->args[0].src == NULL)
				goto error;
			if (build_regex(&c->args[0]))
				goto error;
		}
		if (b != NULL) {
			c->args[1].src = strdup(b);
			if (c->args[1].src == NULL)
				goto error;
			if (build_regex(&c->args[1]))
				goto error;
		}
		c->idx = rs->maxidx++;
		cl->cond = c;
		cl->next = rs->cond[type];
		rs->cond[type] = cl;
	}

	e->type = EXPR_COND;
	e->cond = c;
	e->idx = rs->maxidx++;
	elc->expr = e;
	elc->next = c->expr;
	c->expr = elc;
	mutex_unlock();
	return (e);

error:
	if (elc != NULL)
		free(elc);
	if (e != NULL)
		free(e);
	if (cl != NULL)
		free(cl);
	if (c != NULL) {
		if (!c->args[1].empty)
			regfree(&c->args[1].re);
		if (c->args[1].src != NULL)
			free(c->args[1].src);
		if (!c->args[0].empty)
			regfree(&c->args[0].re);
		if (c->args[0].src != NULL)
			free(c->args[0].src);
		free(c);
	}
	mutex_unlock();
	return (NULL);
}

struct expr *
create_expr(struct ruleset *rs, int type, struct expr *a, struct expr *b)
{
	struct expr *e = NULL;
	struct expr_list *ela = NULL, *elb = NULL;

	mutex_lock();
	e = calloc(1, sizeof(struct expr));
	if (e == NULL)
		goto error;
	if (a != NULL) {
		ela = calloc(1, sizeof(struct expr_list));
		if (ela == NULL)
			goto error;
	}
	if (b != NULL) {
		elb = calloc(1, sizeof(struct expr_list));
		if (elb == NULL)
			goto error;
	}
	e->type = type;
	e->idx = rs->maxidx++;
	if (a != NULL) {
		e->args[0] = a;
		ela->expr = e;
		ela->next = a->expr;
		a->expr = ela;
	}
	if (b != NULL) {
		e->args[1] = b;
		elb->expr = e;
		elb->next = b->expr;
		b->expr = elb;
	}
	mutex_unlock();
	return (e);

error:
	yyerror("create_expr: calloc: %s", strerror(errno));
	if (elb != NULL)
		free(elb);
	if (ela != NULL)
		free(ela);
	if (e != NULL)
		free(e);
	mutex_unlock();
	return (NULL);
}

struct action *
create_action(struct ruleset *rs, int type, const char *msg)
{
	struct action *a = NULL;
	struct action_list *al = NULL;

	mutex_lock();
	a = calloc(1, sizeof(struct action));
	if (a == NULL)
		goto error;
	al = calloc(1, sizeof(struct action_list));
	if (al == NULL)
		goto error;
	a->type = type;
	a->msg = msg == NULL ? NULL : strdup(msg);
	a->idx = rs->maxidx++;
	al->action = a;
	/* tail insert, so actions have same order as file */
	if (rs->action == NULL)
		rs->action = al;
	else {
		struct action_list *t = rs->action;

		while (t->next != NULL)
			t = t->next;
		t->next = al;
	}
	mutex_unlock();
	return (a);

error:
	yyerror("create_action: calloc: %s", strerror(errno));
	if (al != NULL)
		free(al);
	if (a != NULL)
		free(a);
	mutex_unlock();
	return (NULL);
}

struct action *
eval_cond(struct ruleset *rs, int *res, int type,
    const char *a, const char *b)
{
	struct cond_list *cl;
	struct action_list *al;

	mutex_lock();
	for (cl = rs->cond[type]; cl != NULL; cl = cl->next) {
		int r;

		if (res[cl->cond->idx] != VAL_UNDEF)
			continue;
		r = check_cond(cl->cond, a, b);
		if (r < 0) {
			mutex_unlock();
			return (NULL);
		} else if (!r)
			push_cond_result(cl->cond, VAL_TRUE, res);
	}
	for (al = rs->action; al != NULL; al = al->next)
		if (res[al->action->idx] == VAL_TRUE) {
			mutex_unlock();
			return (al->action);
		}
	mutex_unlock();
	return (NULL);
}

struct action *
eval_end(struct ruleset *rs, int *res, int type, int max)
{
	struct cond_list *cl;
	struct action_list *al;

	mutex_lock();
	for (cl = rs->cond[type]; cl != NULL; cl = cl->next)
		if (res[cl->cond->idx] == VAL_UNDEF)
			push_cond_result(cl->cond, VAL_FALSE, res);
	for (al = rs->action; al != NULL; al = al->next)
		if (res[al->action->idx] == VAL_TRUE) {
			mutex_unlock();
			return (al->action);
		}
	for (type = max; type < COND_MAX; ++type)
		for (cl = rs->cond[type]; cl != NULL; cl = cl->next)
			if (res[cl->cond->idx] == VAL_UNDEF) {
				mutex_unlock();
				return (NULL);
			}
	mutex_unlock();
	return (&default_action);
}

void
eval_clear(struct ruleset *rs, int *res, int type)
{
	struct cond_list *cl;
	struct action_list *al;

	mutex_lock();
	for (; type < COND_MAX; ++type)
		for (cl = rs->cond[type]; cl != NULL; cl = cl->next)
			push_cond_result(cl->cond, VAL_UNDEF, res);
	for (al = rs->action; al != NULL; al = al->next)
		res[al->action->idx] = VAL_UNDEF;
	mutex_unlock();
}

static int
check_cond(struct cond *c, const char *a, const char *b)
{
	int i;

	for (i = 0; i < 2; ++i) {
		const char *d = i ? b : a;
		int r;

		if (c->args[i].src == NULL || c->args[i].empty)
			continue;
		if (d == NULL)
			return (-1);
		r = regexec(&c->args[i].re, d, 0, NULL, 0);
		if (r && r != REG_NOMATCH)
			return (-1);
		if ((r == REG_NOMATCH) != c->args[i].not)
			return (1);
	}
	return (0);
}

static void
push_expr_result(struct expr *e, int val, int *res)
{
	struct expr_list *el;

	if (res[e->idx] == val)
		return;
	res[e->idx] = val;
	if (e->action != NULL && val == VAL_TRUE)
		res[e->action->idx] = val;
	for (el = e->expr; el != NULL; el = el->next) {
		struct expr *p = el->expr;

		switch (p->type) {
		case EXPR_AND:
			if (res[p->args[0]->idx] == VAL_TRUE &&
			    res[p->args[1]->idx] == VAL_TRUE)
				push_expr_result(p, VAL_TRUE, res);
			else if (res[p->args[0]->idx] == VAL_FALSE ||
			    res[p->args[1]->idx] == VAL_FALSE)
				push_expr_result(p, VAL_FALSE, res);
			else
				push_expr_result(p, VAL_UNDEF, res);
			break;
		case EXPR_OR:
			if (res[p->args[0]->idx] == VAL_TRUE ||
			    res[p->args[1]->idx] == VAL_TRUE)
				push_expr_result(p, VAL_TRUE, res);
			else if (res[p->args[0]->idx] == VAL_FALSE &&
			    res[p->args[1]->idx] == VAL_FALSE)
				push_expr_result(p, VAL_FALSE, res);
			else
				push_expr_result(p, VAL_UNDEF, res);
			break;
		case EXPR_NOT:
			if (val == VAL_TRUE)
				push_expr_result(p, VAL_FALSE, res);
			else if (val == VAL_FALSE)
				push_expr_result(p, VAL_TRUE, res);
			else
				push_expr_result(p, VAL_UNDEF, res);
			break;
		default:
			break;
		}
	}
}

static void
push_cond_result(struct cond *c, int val, int *res)
{
	struct expr_list *el;

	if (res[c->idx] == val)
		return;
	res[c->idx] = val;
	for (el = c->expr; el != NULL; el = el->next) {
		struct expr *e = el->expr;

		if (e->type != EXPR_COND)
			continue;
		push_expr_result(e, val, res);
	}
}

static int
build_regex(struct cond_arg *a)
{
	char del;
	const char *s = a->src, *t;

	a->empty = 1;
	a->not = 0;
	while (*s == ' ' || *s == '\t')
		s++;
	if (!*s) {
		yyerror("build_regex: empty argument");
		return (1);
	}
	del = *s++;
	t = s;
	while (*s && *s != del)
		s++;
	if (!*s) {
		yyerror("build_regex: missing closing delimiter %s", a->src);
		return (1);
	}
	if (s == t) {
		if (s[1]) {
			yyerror("build_regex: empty expression with flags %s",
			    a->src);
			return (1);
		}
	} else {
		char *u;
		int flags = 0, r;

		u = malloc(s - t + 1);
		if (u == NULL) {
			yyerror("build_regex: malloc: %s", strerror(errno));
			return (1);
		}
		memcpy(u, t, s - t);
		u[s - t] = 0;
		s++;
		while (*s) {
			switch (*s) {
			case 'e':
				flags |= REG_EXTENDED;
				break;
			case 'i':
				flags |= REG_ICASE;
				break;
			case 'n':
				a->not = 1;
				break;
			default:
				yyerror("invalid flag %c in %s", *s, a->src);
				free(u);
				return (1);
			}
			++s;
		}
		if (!(flags & REG_EXTENDED))
			flags |= REG_BASIC;
		r = regcomp(&a->re, u, flags);
		if (r) {
			char e[8192];

			regerror(r, &a->re, e, sizeof(e));
			yyerror("regcomp: %s: %s\n", u, e);
			free(u);
			return (1);
		}
		free(u);
		a->empty = 0;
	}
	return (0);
}

static void
free_expr_list(struct expr_list *el, struct expr *a)
{
	struct expr_list *eln;

	while (el != NULL) {
		struct expr *e = el->expr;

		eln = el->next;
		if (e != NULL) {
			int i, used = 0;

			for (i = 0; i < 2; ++i)
				if (e->args[i] != NULL) {
					if (e->args[i] == a)
						e->args[i] = NULL;
					else
						used = 1;
				}
			if (!used) {
				free_expr_list(e->expr, e);
				free(e);
			}
		}
		free(el);
		el = eln;
	}
}

void
free_ruleset(struct ruleset *rs)
{
	int i;
	struct action_list *al, *aln;

	mutex_lock();
	if (rs == NULL || rs->refcnt) {
		mutex_unlock();
		return;
	}
	for (i = 0; i < COND_MAX; ++i) {
		struct cond_list *cl = rs->cond[i], *cln;

		while (cl != NULL) {
			struct cond *c = cl->cond;

			cln = cl->next;
			if (c != NULL) {
				int j;

				for (j = 0; j < 2; ++j)
					if (c->args[j].src != NULL) {
						free(c->args[j].src);
						if (!c->args[j].empty)
							regfree(&c->args[j].re);
					}
				free_expr_list(c->expr, NULL);
				free(c);
			}
			free(cl);
			cl = cln;
		}
	}
	al = rs->action;
	while (al != NULL) {
		struct action *a = al->action;

		aln = al->next;
		if (a != NULL) {
			if (a->msg != NULL)
				free(a->msg);
			free(a);
		}
		free(al);
		al = aln;
	}
	free(rs);
	mutex_unlock();
}


eval.h
/* $Id: eval.h,v 1.1.1.1 2007/01/11 15:49:52 dhartmei Exp $ */

/*
 * Copyright (c) 2004-2006 Daniel Hartmeier
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *    - Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    - Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * ChangeLog:
 * 2022/04/17  Add function prototype of geoip.c
 */

#ifndef _EVAL_H_
#define _EVAL_H_

#include <regex.h>

enum { VAL_UNDEF=0, VAL_TRUE, VAL_FALSE };
enum { COND_MACRO, COND_CONNECT, COND_COUNTRY, COND_HELO, COND_ENVFROM, COND_ENVRCPT,
    COND_HEADER, COND_BODY, COND_MAX };
enum { EXPR_AND, EXPR_OR, EXPR_NOT, EXPR_COND };
enum { ACTION_REJECT, ACTION_TEMPFAIL, ACTION_QUARANTINE,
    ACTION_DISCARD, ACTION_ACCEPT };

struct expr;

struct cond {
	struct cond_arg {
		char	*src;
		int	 empty;
		int	 not;
		regex_t	 re;
	}			 args[2];
	struct expr_list	*expr;
	unsigned		 idx;
};

struct cond_list {
	struct cond		*cond;
	struct cond_list	*next;
};

struct action;

struct expr {
	int			 type;
	struct expr		*args[2];
	struct cond		*cond;
	struct action		*action;
	struct expr_list	*expr;
	unsigned		 idx;
};

struct expr_list {
	struct expr		*expr;
	struct expr_list	*next;
};

struct action {
	int			 type;
	char			*msg;
	unsigned		 idx;
};

struct action_list {
	struct action		*action;
	struct action_list	*next;
};

struct ruleset {
	struct cond_list	*cond[COND_MAX];
	struct action_list	*action;
	unsigned		 maxidx;
	unsigned		 refcnt;
};

int		 eval_init(int);
struct ruleset	*create_ruleset(void);
struct expr	*create_cond(struct ruleset *, int, const char *,
		    const char *);
struct expr	*create_expr(struct ruleset *, int, struct expr *,
		    struct expr *);
struct action	*create_action(struct ruleset *, int, const char *);
struct action	*eval_cond(struct ruleset *, int *, int,
		    const char *, const char *);
struct action	*eval_end(struct ruleset *, int *, int, int);
void		 eval_clear(struct ruleset *, int *, int);
void		 free_ruleset(struct ruleset *);

/* Function prototype of geoip.c */

extern  void init_geoip( ) ;
extern  int  get_CountryCode( const char *, char[3] ) ;   /* Return country code into 2nd. param. */
extern  int  check_geoipEnabled( ) ;                      /* 0: Enabled / 1: Disabled */

extern  void reset_geoip( ) ;
extern  void set_ipv4file( const char * ) ;
extern  void set_ipv6file( const char * ) ;
extern  void set_geoipCashMax( const char * ) ;
extern  void set_geoipCashKeep( const char * ) ;
extern  void set_geoipDebugLog( const char * ) ;

#endif


strlcat.c
/* $Id: strlcat.c,v 1.1 2010/08/30 12:51:49 dhartmei Exp $ */
/*	$OpenBSD: strlcat.c,v 1.12 2005/03/30 20:13:52 otto Exp $	*/

/*
 * Copyright (c) 1998 Todd C. Miller <Todd.Miller@courtesan.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <string.h>

/*
 * Appends src to string dst of size siz (unlike strncat, siz is the
 * full size of dst, not space left).  At most siz-1 characters
 * will be copied.  Always NUL terminates (unless siz <= strlen(dst)).
 * Returns strlen(src) + MIN(siz, strlen(initial dst)).
 * If retval >= siz, truncation occurred.
 */
size_t
strlcat(char *dst, const char *src, size_t siz)
{
	char *d = dst;
	const char *s = src;
	size_t n = siz;
	size_t dlen;

	/* Find the end of dst and adjust bytes left but don't go past end */
	while (n-- != 0 && *d != '\0')
		d++;
	dlen = d - dst;
	n = siz - dlen;

	if (n == 0)
		return(dlen + strlen(s));
	while (*s != '\0') {
		if (n != 1) {
			*d++ = *s;
			n--;
		}
		s++;
	}
	*d = '\0';

	return(dlen + (s - src));	/* count does not include NUL */
}


strlcpy.c
/* $Id: strlcpy.c,v 1.1.1.1 2007/01/11 15:49:52 dhartmei Exp $ */
/*	$OpenBSD: strlcpy.c,v 1.8 2003/06/17 21:56:24 millert Exp $	*/

/*
 * Copyright (c) 1998 Todd C. Miller <Todd.Miller@courtesan.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

static const char rcsid[] = "$Id: strlcpy.c,v 1.1.1.1 2007/01/11 15:49:52 dhartmei Exp $";

#include <sys/types.h>
#include <string.h>

/*
 * Copy src to string dst of size siz.  At most siz-1 characters
 * will be copied.  Always NUL terminates (unless siz == 0).
 * Returns strlen(src); if retval >= siz, truncation occurred.
 */
size_t
strlcpy(char *dst, const char *src, size_t siz)
{
	register char *d = dst;
	register const char *s = src;
	register size_t n = siz;

	/* Copy as many bytes as will fit */
	if (n != 0 && --n != 0) {
		do {
			if ((*d++ = *s++) == 0)
				break;
		} while (--n != 0);
	}

	/* Not enough room in dst, add NUL and traverse rest of src */
	if (n == 0) {
		if (siz != 0)
			*d = '\0';		/* NUL-terminate dst */
		while (*s++)
			;
	}

	return(s - src - 1);	/* count does not include NUL */
}
